File: AOC_2022/Day 7/7.py
import re

file = open("input.txt", "r")

lines = file.read().splitlines()

arr = []
total_size = 0
result = 0
i = 2

def helper():
    s = 0
    global i
    while i < len(lines):
        if "$ ls" == lines[i] or "dir" in lines[i]:
            i += 1
            continue
        else:
            if bool(re.search(r'\d', lines[i])):
                s += int(re.sub("\D", "", lines[i]))
            else:
                if "$ cd .." == lines[i]:
                    if s < 100000:
                        global result
                        result += s
                    arr.append(s)
                    return s
                else:
                    i += 1
                    s += helper()
        i += 1
    return s


while i < len(lines):
    if bool(re.search(r'\d', lines[i])):
        total_size += int(re.sub("\D", "", lines[i]))
    elif "$ cd " not in lines[i]:
        i += 1
        continue
    else:
        i += 1
        x = helper()
        total_size += x
        #arr.append(x)
    i += 1

print("The Result for the 1st star is: " + str(result))

#print(arr)

print("The Result for the 2nd star is: " +
      str(min(n for n in arr if n > (total_size - 40000000))))

File: AOC_2022/Day 7/7.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstring>
#include <cctype>
#include <vector>
#include <algorithm>
using namespace std;

#define ll long long int

ll result = 0;
ll total_size_possible = 40000000, total_size = 0;

ifstream in("input.txt");
string line;

bool ignore(string str){
    return ((line.find("$ ls") != string::npos) || (line.find("dir") == 0));
}

vector<int> vec;

ll helper(){
    ll temp = 0;
    while (getline(in, line))
    {
        if(ignore(line)) continue;
        else{
            if(any_of(line.begin(), line.end(), ::isdigit)){
                ll x = 0;
                for(int i = 0; i < line.size(); i++){
                    if(isdigit(line[i])){
                        x = x*10 + (line[i] - '0');
                    }
                }
                temp += x;
            }
            else{
                if(line == "$ cd .."){
                    if(temp < 1e5) result += temp;
                    vec.push_back(temp);
                    return temp;
                }
                else{
                    temp += helper();
                }
            }
        }
    }
    return temp;
}

int main()
{
    ll tmp;
    getline(in, line);
    getline(in, line);
    while (getline(in, line))
    {
        if(any_of(line.begin(), line.end(), ::isdigit)){
            ll x = 0;
            for(int i = 0; i < line.size(); i++){
                if(isdigit(line[i])){
                    x = x*10 + (line[i] - '0');
                }
            }
            total_size += x;
        }
        else if(line.find("$ cd ") == string::npos)
        {
            continue;
        }
        else{
            tmp = helper();
            total_size += tmp;
            vec.push_back(tmp);
        }
    }
    cout<<result<<"\n";
    int to_remove = total_size - total_size_possible;
    sort(vec.begin(), vec.end());
    //for (auto i: vec) cout<<i<<" ";
    cout<<"\n";
    cout<<vec[lower_bound(vec.begin(), vec.end(), to_remove) - vec.begin()]<<"\n";
    return 0;
}
